huaweicloud/resource_huaweicloud_networking_secgroup_rule.go

```diff
@@ -123,122 +120,138 @@ func ResourceNetworkingSecGroupRule() *schema.Resource {
-		if d.Get("port_range_max").(int) == val.(int) {
-			return strconv.Itoa(val.(int))
+func resourceNetworkingSecGroupRuleCreate(ctx context.Context, d *schema.ResourceData,
+	meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	region := GetRegion(d, config)
+
+	if _, ok := d.GetOk("ports"); ok {
+		v3Client, err := config.NetworkingV3Client(region)
+		if err != nil {
+			return fmtp.DiagErrorf("Error creating HuaweiCloud networking v3 client: %s", err)
+		}
+		err = createRuleByV3Client(v3Client, d, config)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	} else {
+		v1Client, err := config.SecurityGroupV1Client(region)
+		if err != nil {
+			return fmtp.DiagErrorf("Error creating HuaweiCloud networking v1 client: %s", err)
+		}
+		err = createRuleByV1Client(v1Client, d, config)
+		if err != nil {
+			return diag.FromErr(err)
 		}
-		return fmt.Sprintf("%d-%d", val.(int), d.Get("port_range_max").(int))
 	}
-	return d.Get("ports").(string)
+
+	return resourceNetworkingSecGroupRuleRead(ctx, d, meta)
 }
 
 func resourceNetworkingSecGroupRuleRead(_ context.Context, d *schema.ResourceData,
 	meta interface{}) diag.Diagnostics {
-	logp.Printf("[DEBUG] Retrieve information about security group rule: %s", d.Id())
-
 	config := meta.(*config.Config)
-	client, err := config.NetworkingV3Client(GetRegion(d, config))
+	region := GetRegion(d, config)
+
+	v1Client, err := config.SecurityGroupV1Client(region)
 	if err != nil {
-		return fmtp.DiagErrorf("Error creating HuaweiCloud networking client: %s", err)
+		return fmtp.DiagErrorf("Error creating HuaweiCloud networking v1 client: %s", err)
+	}
+	v3Client, err := config.NetworkingV3Client(GetRegion(d, config))
+	if err != nil {
+		return fmtp.DiagErrorf("Error creating HuaweiCloud networking v3 client: %s", err)
 	}
 
-	resp, err := rules.Get(client, d.Id())
-
+	resp, err := v1Rules.Get(v1Client, d.Id())
 	if err != nil {
+		logp.Printf("[DEBUG] Unable to find the specified Security group rule (%s).", d.Id())
 		return common.CheckDeletedDiag(d, err, "HuaweiCloud Security Group Rule")
 	}
 
 	mErr := multierror.Append(nil,
+		d.Set("region", GetRegion(d, config)),
 		d.Set("direction", resp.Direction),
 		d.Set("description", resp.Description),
 		d.Set("ethertype", resp.Ethertype),
 		d.Set("protocol", resp.Protocol),
 		d.Set("remote_group_id", resp.RemoteGroupId),
 		d.Set("remote_ip_prefix", resp.RemoteIpPrefix),
 		d.Set("security_group_id", resp.SecurityGroupId),
-		d.Set("region", GetRegion(d, config)),
-		setSecurityGroupRulePorts(d, resp.MultiPort),
+		d.Set("port_range_min", resp.PortRangeMin),
+		d.Set("port_range_max", resp.PortRangeMax),
 	)
-	if mErr.ErrorOrNil() != nil {
-		return diag.FromErr(mErr)
-	}
 
-	return nil
-}
-
-func setSecurityGroupRulePorts(d *schema.ResourceData, portRange string) error {
-	if portRange == "" {
-		return nil
+	rule, err := v3Rules.Get(v3Client, d.Id())
+	if _, ok := err.(golangsdk.ErrDefault404); ok {
```

Comment:
what if the error is not 404?

Reply:
Since the V1 version of the query was previously performed, if the resource does not exist, the function process will not reach this step, so only the situation where the API does not exist is left.